projects/RabbitMQ.Client/CreateChannelOptions.cs

```diff
@@ -53,7 +53,7 @@ public sealed class CreateChannelOptions
         /// <see cref="IChannel.GetNextPublishSequenceNumberAsync(System.Threading.CancellationToken)"/> to allow correlation
         /// of the response with the correct message.
         /// </summary>
-        public bool PublisherConfirmationsEnabled { get; set; } = false;
+        public readonly bool PublisherConfirmationsEnabled = false;
```

Comment:
See, that's the thing. Now these become fields because we don't have pollyfill that would have allowed us to have init only members. 

I feel like the other changes that we made even though it was already RC have been kind of justifiable but this breaking change seems to now almost go into slightly esoteric territory for little value.

If you feel it is worth doing this change and other such classes in the client are also using public fields, then fair enough. Otherwise, for a public options type `init` only members of a class or a record type that allows "cloning" with the `with` expression is far more intuitive.

In the Azure SDK the options type are mostly classes with getters and setters but they predate init only members.

Reply:
I just don't want someone to hold onto a `CreateChannelOptions` reference and unknowingly update it, which could screw up channel recovery, since a reference is held to the options in `AutorecoveringChannel`. It's unlikely, of course.

This is why I created the `ChannelOptions` type, because it "froze" the options when created.

Reply:
So, I'm going to either proceed with this change or bring back `ChannelOptions` 😄

Reply:
`init` only fields do not guarantee that.

You can use a `struct`. That will always be copied, thus, not a good idea if they are big.

One strategy I use some times is using a context type (think `HttpContext`) that keeps the state of the operation. 

Sorry, I'm on the phone and looking at the code is hard.

Reply:
> `init` only fields do not guarantee that.
> 
> You can use a `struct`. That will always be copied, thus, not a good idea if they are big.
> 
> One strategy I use some times is using a context type (think `HttpContext`) that keeps the state of the operation.
> 
> Sorry, I'm on the phone and looking at the code is hard.

![Screenshot_2024-10-24-07-32-08-340_com duckduckgo mobile android-edit](https://github.com/user-attachments/assets/45a8cd55-54b2-4cee-a59b-2fdb19ceca3f)

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init

Reply:
> I just don't want someone to hold onto a `CreateChannelOptions` reference and unknowingly update it, which could screw up channel recovery (or, at least, change the channel that is created during recovery), since a reference is held to the options in `AutorecoveringChannel`. It's unlikely, of course.
> 
> This is why I created the `ChannelOptions` type, because it "froze" the options when created.

That's fine and a good idea. Init only would give you that or simply "shallow cloning" the same object no?

Reply:
The other problem is that the more complicated and non standard those option types become the less likely it is you can configure those with Microsoft.Extension.Options. But the rate limiter there with only a non-default ctor makes that anyway already a bit harder to maybe that's a mute point

Reply:
> So, I'm going to either proceed with this change or bring back `ChannelOptions` 😄

My comment on the other PR was simply curiosity. You could have answered it with

> I just don't want someone to hold onto a CreateChannelOptions reference and unknowingly update it, which could screw up channel recovery (or, at least, change the channel that is created during recovery), since a reference is held to the options in AutorecoveringChannel.

😂🤣

Now looking at it and given the current constraints with init only member not being available or us requiring some extra adjustment logic internally maybe that extra internal type is justified

Reply:
> > `init` only fields do not guarantee that.
> > You can use a `struct`. That will always be copied, thus, not a good idea if they are big.
> > One strategy I use some times is using a context type (think `HttpContext`) that keeps the state of the operation.
> > Sorry, I'm on the phone and looking at the code is hard.
> 
> ![Screenshot_2024-10-24-07-32-08-340_com duckduckgo mobile android-edit](https://private-user-images.githubusercontent.com/174258/379627854-45a8cd55-54b2-4cee-a59b-2fdb19ceca3f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NTQ3MDYsIm5iZiI6MTcyOTc1NDQwNiwicGF0aCI6Ii8xNzQyNTgvMzc5NjI3ODU0LTQ1YThjZDU1LTU0YjItNGNlZS1hNTliLTJmZGIxOWNlY2EzZi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQwNzIwMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Mjk0ZmVlY2QyYTAwNjkwYjJiMDVlNjRlZTE1OWQ5MjVlZjBkZDg5ZGYzZGNjZTEwZTA4ZjBhZTlhNjVjYjQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.rSy8qx1riAh_27JLDinhJnxO7lacUuX7Z_tPiTtclRc)
> 
> https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init

`init` only members is a C# language feature that might not be implemented or respected by other .NET languages, like Visual Basic.

If you look at the underlying implementation, you'll see that there's a `set_XXX` method, not a `init_XXX` method. [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYUIG9DSHyyMcAGUgZVJtIHMBTBAG5SsKMNIBfQpKA==)

Unless it's clear that this library is only supported when consumed by C# code, you can't rely on properties not being changed after the object being handed to you.

Reply:
The IL field has the init only and the set_XXX is passed into modreg which takes care of that.

> "The primary languages for .NET (C#, F# and VB) will all be updated to recognize these init accessors. Hence the only realistic scenario here is when a C# 9 compiler emits init properties and they are seen by an older toolset such as C# 8, VB 15, etc ... C# 8. That is the trade off to consider and weigh against binary compatibility." 

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/init

Last issue in VB was fixed in 16.9.4 https://github.com/dotnet/roslyn/pull/50414 

 I'm not sure there is much value in further discussing this honestly because it is supported and works as specified. Anything else potentially but quite unlikely remaining is pure super edge case scenario out of spec so 🤷